designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,784 @@
+
+#### Extending From `eslint:recommended` and `eslint:all`
+
+Both `eslint:recommended` and `eslint:all` can be represented as strings in a config array. For example:
+
+```js
+exports.config = [
+    "eslint:recommended",
+    require("eslint-config-standard"),
+    require("@me/eslint-config"),
+    {
+        files: ["*.js"],
+        rules: {
+            semi: ["error", "always"]
+        }
+    }
+];
+```
+
+This config first extends `eslint:recommended` and then continues on to extend other configs.
+
+#### Setting the Name of Shareable Configs
+
+For shareable configs, specifying a `name` property for each config they export helps ESLint to output more useful error messages if there is a problem. The `name` property is a string that will be used to identify configs to help users resolve problems. For example, if you are creating `eslint-config-example`, then you can specify a `name` property to reflect that:
+
+```js
+module.exports = {
+    name: "eslint-config-example",
+
+    // other info here
+};
+```
+
+It's recommended that the shareable config provide a unique name for each config that is exported.
+
+#### Disambiguating Shareable Configs With Common Dependencies
+
+Today, shareable configs that depend on plugin rules must specify the plugin as a peer dependency and then either provide a script to install those dependencies or ask the user to install them manually.
+
+With this design, shareable configs can specify plugins as direct dependencies that will be automatically installed with the shareable config, improving the user experience of complex shareable configs. This means its possible for multiple shareable configs to depend on the same plugin and, in theory, depend on different versions of the same plugin. In general, npm will handle this directly, installing the correct plugin version at the correct level for each shareable config to `require()`. For example, suppose there are two shareable configs, `eslint-config-a` and `eslint-config-b` that both rely on the `eslint-plugin-example` plugin, but the former relies on 1.0.0 and the latter relies on 2.0.0. npm will install those plugins like this:
+
+```
+your-project
+├── eslint.config.js
+└── node_modules
+  ├── eslint
+  ├── eslint-config-a
+  | └── node_modules
+  |   └── eslint-plugin-example@1.0.0
+  └── eslint-config-b
+    └── node_modules
+      └── eslint-plugin-example@2.0.0
+```
+
+The problem comes when the shareable configs try to use the default namespace of `eslint-plugin-example` for its rules, such as:
+
+```js
+module.exports = {
+    ruledefs: {
+        ...require("eslint-plugin-example").ruledefs
+    }
+};
+```
+
+If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown because the `ruledefs` namespace "example" can only be assigned once.
```

Comment:
> For example, if the `eslint-plugin-node` plugin exports two configs, both of which use `eslint-plugin-node`'s rules, then the user won't be able to extend both of them because they will both have a `node` ruledef containing `eslint-plugin-node`'s rules, even though both configs were created by the same author and intended to share the same rules.

Do you know of an example where this happens today?

My opinion is that it would be up to the plugin to provide documentation about the correct way to use its exported configs.

Reply:
> Do you know of an example where this happens today?

Yes, I think it's pretty common. For example, [`eslint-plugin-eslint-plugin`](https://github.com/not-an-aardvark/eslint-plugin-eslint-plugin/blob/9bab97423656bb40a4cd199034c5a167e89b822e/lib/index.js) exports `rules` and `tests` configs which enable different subsets of the plugin's rules.